src/execution/operator/scan/physical_table_scan.cpp

```diff
@@ -73,7 +111,15 @@ SourceResultType PhysicalTableScan::GetData(ExecutionContext &context, DataChunk
 	TableFunctionInput data(bind_data.get(), state.local_state.get(), gstate.global_state.get());
 	function.function(context.client, data, chunk);
 
-	return chunk.size() == 0 ? SourceResultType::FINISHED : SourceResultType::HAVE_MORE_OUTPUT;
+	if (chunk.size() == 0) {
```

Comment:
For regular table functions, since they only operate on constant values, `WITH ORDINALITY` is always equivalent to `row_number() over ()`, i.e.:

```sql
SELECT * FROM read_csv('file.csv') WITH ORDINALITY;
```

Is equivalent to:

```sql
SELECT * FROM (SELECT *, row_number() OVER () AS ordinality FROM read_csv('file.csv'));
```
As such, I think an easier/more robust way of generating the ordinality column for a `PhysicalTableScan` is to push a `Window` operator during physical planning time. Then all regular `TableFunctions` will immediately support the `WITH ORDINALITY` clause and it only needs to be implemented for table in-out functions.

We can then also simplify the code, e.g. the `MultiFileReader` does not need to be modified, and perhaps we can even remove `supports_ordinality` from the `TableFunction` and support it generically for all table functions (both regular and table in-out functions).

Reply:
This works fine for TableFunctions (see latest commit), but where it falls apart is TableInOutFunctions. 
More specifically, the InOutFunction (well, Unnest(), since it's the only function right now) can not differentiate between seperate inputs in this approach. So the query 
`SELECT * FROM UNNEST((SELECT [1,2,3] UNION ALL SELECT [4,5])) with ordinality;`
yields

1	1
2	2
3	3
4	4
5	5

instead of 

1	1
2	2
3	3
4	1
5	2

So, for InOutFunctions, I have to use the old approach (of the previous commits before _[da1bb36]_) to differentiate the inputs. 
The problem is that right now, InOut and regular TableFunctions share the same BindData. The only currently implemented InOut-Function is Unnest(), which has its own BindData inherting from FunctionData directly (instead of TableFunctionData). This is because Unnest() behaves differently when used directly within a SELECT clause. With the old approach, this was the only reason I had to include my additional fields in the general FunctionData instead of TableFunctionData, as it is difficult and unhandy to cast to UnnestBindData in some specific cases. The problem above would be easily fixable (for all future TableInOutFunctions) if besides TableFunctionData, there would also be TableInOutFunctionData. Another option would be for UnnestBindData to inherit from TableFunctionData instead of FunctionData.

My proposal is to implement TableFunctionInOutData, what are your thoughts on this?

Reply:
That sounds like a good idea.

Reply:
^ This would also make it easier to turn functions like the range() function into an InOutFunction that can accept rows of a table-input as arguments, which is something I want to get my hands on next. 

Are there any problems with this approach that I'm not seeing?

Reply:
I don't think so - it would be great to get other functions migrated over to the table in-out function framework

Reply:
I've picked up turning `range/generate_series` into an InOutFunction in https://github.com/duckdb/duckdb/pull/12431 - I don't think it conflicts too much with this PR but there might be some conflicts that pop up.